### PR TITLE
Fixes #206 by adding context for false branch in IfInState TypeChecker

### LINKED ISSUE #206: Make dynamic state check type reasoning smarter
```
 transaction t4(LightSwitch@Owned s) {
        if (s in On) {
             s.turnOff();
        }

        [s @ Off];
}
```
The assertion above currently fails because the type checker isn't smart enough. Let's make it succeed (assuming that the only possible states are On and Off).

### PR DESCRIPTION
**Pull request by YBthebest**
We add a different context that expands the Owner states into all possible states and removes the one state we checked for in the true branch context. This fixes #206

### CODE REVIEW

File: src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala [hunk 1]
```diff
@@ -1468,17 +1488,19 @@ private def checkStatement(
                                     np.permission match {
                                         case Owned() =>
                                             val newType = StateType(np.contractName, Set(state._1), np.isRemote)
-                                            contextPrime.updated(x, newType).updatedMakingVariableVal(x)
-                                        case Unowned() => contextPrime
+                                            val typeFalse = StateType(np.contractName, allStates - state._1, np.isRemote)
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala]: This is close, but what if this is a StateType (not just any old NonPrimitiveType)? Then we know something more specific about the possible states than just that it's one of allStates.

File: resources/tests/type_checker_tests/InState.obs [hunk 2]
```diff
@@ -51,13 +51,13 @@ main contract InState {
...
 
-        // TODO: Make the type checker smart enough for this assertion to pass.
-        // [s @ Off];
+        [s @ Off];
 
         if (s in Off) {
              s.turnOn();
         }
 
+        // ERROR: s is On now, we know this specifically
```
Comment (mcoblenz) [inline resources/tests/type_checker_tests/InState.obs]: Need to update the unit tests to expect this error.

File: src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala (hunk 1 above)
Comment (YBthebest) [reply src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala]: @mcoblenz I put allStates - state._1, isn't that what you are referring too?

File: resources/tests/type_checker_tests/InState.obs (hunk 2 above)
Comment (YBthebest) [reply resources/tests/type_checker_tests/InState.obs]: Where can I find the unit tests?

File: src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala (hunk 1 above)
Comment (mcoblenz) [reply src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala]: But allStates is all the states in the contract. If the current type is a particular set of states, you should use that instead.

File: resources/tests/type_checker_tests/InState.obs (hunk 2 above)
Comment (mcoblenz) [reply resources/tests/type_checker_tests/InState.obs]: Look at TypeCheckerTests.scala. Or run sbt test and you'll see what fails.

**Pull request merged**